### PR TITLE
waterfall rules

### PR DESCRIPTION
**Pull request by IhorEhorovCoax**

### CODE REVIEW

File: lib/passpartu/verify.rb [hunk 1]
```diff
@@ -36,13 +40,46 @@ def role_ignore?
...
+        @result = hash_include_keys?(policy_hash, keys) ? policy_hash.dig(*keys) : nil
+        break if @result || keys.empty?
+      end
+    end
+
+    def hash_include_keys?(hash, keys)
+      flatten_hash(hash).keys.any? { |k| k.include?(keys.join('_')) }
+    end
+
+    def flatten_hash(hash)
```
Comment (OrestF) [inline lib/passpartu/verify.rb]: This method is redundant, U think!

File: lib/passpartu/check_policy.rb [hunk 2]
```diff
@@ -0,0 +1,59 @@
+module Passpartu
+  class CheckPolicy
+    CRUD_KEY = 'crud'.freeze
+
+    def initialize(keys)
+      @hash = Passpartu.policy
```
Comment (OrestF) [inline lib/passpartu/check_policy.rb]: bad name

File: lib/passpartu/check_policy.rb [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+      @keys = keys
```
Comment (OrestF) [inline lib/passpartu/check_policy.rb]: redundand

File: lib/passpartu/check_policy.rb [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+      @waterfall_rules = Passpartu.config.waterfall_rules
```
Comment (OrestF) [inline lib/passpartu/check_policy.rb]: redundant

File: lib/passpartu/check_policy.rb [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+      check_policy
+      check_crud if policy_missed? && last_key_crud?
+      result
+    end
+
+    private
+
+    attr_reader :hash, :keys, :waterfall_rules, :result
+
+    def check_policy
```
Comment (OrestF) [inline lib/passpartu/check_policy.rb]: name it something with `..waterfall..`

File: lib/passpartu/check_policy.rb [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+      loop_hash = hash.dup
+      keys.each_with_index do |key, index|
+        if loop_hash[key].is_a? Hash
+          loop_hash = loop_hash[key]
+          next
+        elsif waterfall_rules || last?(index)
```
Comment (OrestF) [inline lib/passpartu/check_policy.rb]: check this config outside method logic. Otherwords - do not check waterfall if it's turned off

File: lib/passpartu/check_policy.rb [hunk 7]
```diff
@@ -0,0 +1,59 @@
+module Passpartu
+  class CheckPolicy
```
Comment (OrestF) [inline lib/passpartu/check_policy.rb]: I meant, move waterfall logic into separate module, but not the whole verification logic. `passaprtu/verify` it's a core, and `passpartu/waterfall_rules` should be a component!

File: lib/passpartu/verify.rb [hunk 8]
```diff
@@ -37,7 +45,21 @@ def role_ignore?
     end
 
     def check_policy
-      @result = Passpartu.policy.dig(role, *keys)
+      @result = waterfall_rules? ? check_waterfall_rules : simple_check
+    end
+
+    def simple_check
```
Comment (OrestF) [inline lib/passpartu/verify.rb]: What is it?
and what the difference between  `check_waterfall_rules`

Comment (IhorEhorovCoax) [reply lib/passpartu/verify.rb]: simple_check need when waterfall_rules is false, so it don't check waterfall

**Pull request closed without merging**